File: app/src/test/java/org/alkaline/taskbrain/ui/currentnote/util/CompletedLineUtilsTest.kt
package org.alkaline.taskbrain.ui.currentnote.util

import org.alkaline.taskbrain.ui.currentnote.util.CompletedLineUtils.DisplayItem
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class CompletedLineUtilsTest {

    // ==================== isCheckedCheckbox ====================

    @Test
    fun `isCheckedCheckbox returns true for checked line`() {
        assertTrue(CompletedLineUtils.isCheckedCheckbox("☑ Done"))
    }

    @Test
    fun `isCheckedCheckbox returns true for indented checked line`() {
        assertTrue(CompletedLineUtils.isCheckedCheckbox("\t\t☑ Done"))
    }

    @Test
    fun `isCheckedCheckbox returns false for unchecked line`() {
        assertFalse(CompletedLineUtils.isCheckedCheckbox("☐ Todo"))
    }

    @Test
    fun `isCheckedCheckbox returns false for plain line`() {
        assertFalse(CompletedLineUtils.isCheckedCheckbox("Plain text"))
    }

    @Test
    fun `isCheckedCheckbox returns false for empty line`() {
        assertFalse(CompletedLineUtils.isCheckedCheckbox(""))
    }

    // ==================== computeHiddenIndices ====================

    @Test
    fun `showCompleted true returns empty set`() {
        val lines = listOf("Title", "☑ Done", "☐ Active")
        assertEquals(emptySet<Int>(), CompletedLineUtils.computeHiddenIndices(lines, true))
    }

    @Test
    fun `no checked lines returns empty set`() {
        val lines = listOf("Title", "☐ Active", "• Bullet")
        assertEquals(emptySet<Int>(), CompletedLineUtils.computeHiddenIndices(lines, false))
    }

    @Test
    fun `checked line is hidden`() {
        val lines = listOf("Title", "☑ Done", "☐ Active")
        assertEquals(setOf(1), CompletedLineUtils.computeHiddenIndices(lines, false))
    }

    @Test
    fun `checked line with children hides entire subtree`() {
        val lines = listOf("Title", "☑ Done", "\tChild 1", "\tChild 2", "☐ Active")
        assertEquals(setOf(1, 2, 3), CompletedLineUtils.computeHiddenIndices(lines, false))
    }

    @Test
    fun `nested checked hides subtree`() {
        val lines = listOf("Title", "Parent", "\t☑ Done", "\t\tGrandchild", "\tSibling")
        assertEquals(setOf(2, 3), CompletedLineUtils.computeHiddenIndices(lines, false))
    }

    @Test
    fun `title line is never hidden even if checked`() {
        val lines = listOf("☑ Checked Title", "☐ Active")
        assertEquals(emptySet<Int>(), CompletedLineUtils.computeHiddenIndices(lines, false))
    }

    @Test
    fun `multiple checked lines with children`() {
        val lines = listOf("Title", "☑ Done1", "\tChild1", "☑ Done2", "\tChild2", "☐ Active")
        assertEquals(setOf(1, 2, 3, 4), CompletedLineUtils.computeHiddenIndices(lines, false))
    }

    // ==================== computeDisplayItems ====================

    @Test
    fun `showCompleted true returns all visible lines`() {
        val lines = listOf("Title", "☑ Done", "☐ Active")
        val items = CompletedLineUtils.computeDisplayItems(lines, true)
        assertEquals(3, items.size)
        assertEquals(DisplayItem.VisibleLine(0), items[0])
        assertEquals(DisplayItem.VisibleLine(1), items[1])
        assertEquals(DisplayItem.VisibleLine(2), items[2])
    }

    @Test
    fun `hidden lines produce placeholder with correct count`() {
        val lines = listOf("Title", "☑ Done1", "☑ Done2", "☐ Active")
        val items = CompletedLineUtils.computeDisplayItems(lines, false)
        assertEquals(3, items.size)
        assertEquals(DisplayItem.VisibleLine(0), items[0])
        assertEquals(DisplayItem.CompletedPlaceholder(2, 0, 1), items[1])
        assertEquals(DisplayItem.VisibleLine(3), items[2])
    }

    @Test
    fun `placeholder counts only top-level hidden items`() {
        val lines = listOf("Title", "☑ Done", "\tChild1", "\tChild2", "☐ Active")
        val items = CompletedLineUtils.computeDisplayItems(lines, false)
        assertEquals(3, items.size)
        assertEquals(DisplayItem.VisibleLine(0), items[0])
        // 1 top-level checked item at indent 0, 2 children at indent 1
        assertEquals(DisplayItem.CompletedPlaceholder(1, 0, 1), items[1])
        assertEquals(DisplayItem.VisibleLine(4), items[2])
    }

    @Test
    fun `indented placeholder has correct indent level`() {
        val lines = listOf("Title", "Parent", "\t☑ Done", "\t☐ Active")
        val items = CompletedLineUtils.computeDisplayItems(lines, false)
        assertEquals(4, items.size)
        assertEquals(DisplayItem.VisibleLine(0), items[0])
        assertEquals(DisplayItem.VisibleLine(1), items[1])
        assertEquals(DisplayItem.CompletedPlaceholder(1, 1, 2), items[2])
        assertEquals(DisplayItem.VisibleLine(3), items[3])
    }

    // ==================== sortCompletedToBottom ====================

    @Test
    fun `empty input returns empty`() {
        assertEquals(emptyList<String>(), CompletedLineUtils.sortCompletedToBottom(emptyList()))
    }

    @Test
    fun `single line returns unchanged`() {
        assertEquals(listOf("Title"), CompletedLineUtils.sortCompletedToBottom(listOf("Title")))
    }

    @Test
    fun `no checked lines returns unchanged`() {
        val lines = listOf("Title", "☐ A", "☐ B")
        assertEquals(lines, CompletedLineUtils.sortCompletedToBottom(lines))
    }

    @Test
    fun `checked lines move to bottom`() {
        val lines = listOf("Title", "☑ Done", "☐ Active")
        val expected = listOf("Title", "☐ Active", "☑ Done")
        assertEquals(expected, CompletedLineUtils.sortCompletedToBottom(lines))
    }

    @Test
    fun `multiple checked and unchecked sort correctly`() {
        val lines = listOf("Title", "☑ Done A", "☐ Active B", "☐ Active C", "☑ Done D")
        val expected = listOf("Title", "☐ Active B", "☐ Active C", "☑ Done A", "☑ Done D")
        assertEquals(expected, CompletedLineUtils.sortCompletedToBottom(lines))
    }

    @Test
    fun `checked subtrees stay intact`() {
        val lines = listOf("Title", "☑ Done", "\tChild", "☐ Active")
        val expected = listOf("Title", "☐ Active", "☑ Done", "\tChild")
        assertEquals(expected, CompletedLineUtils.sortCompletedToBottom(lines))
    }

    @Test
    fun `nested levels sort independently`() {
        val lines = listOf(
            "Title",
            "Parent",
            "\t☑ Done child",
            "\t☐ Active child",
        )
        val expected = listOf(
            "Title",
            "Parent",
            "\t☐ Active child",
            "\t☑ Done child",
        )
        assertEquals(expected, CompletedLineUtils.sortCompletedToBottom(lines))
    }

    @Test
    fun `empty lines act as barriers`() {
        val lines = listOf(
            "Title",
            "☑ Done A",
            "☐ Active B",
            "",
            "☑ Done C",
            "☐ Active D",
        )
        val expected = listOf(
            "Title",
            "☐ Active B",
            "☑ Done A",
            "",
            "☐ Active D",
            "☑ Done C",
        )
        assertEquals(expected, CompletedLineUtils.sortCompletedToBottom(lines))
    }

    @Test
    fun `already sorted is idempotent`() {
        val lines = listOf("Title", "☐ Active", "☑ Done")
        assertEquals(lines, CompletedLineUtils.sortCompletedToBottom(lines))
    }

    @Test
    fun `deep nesting sorts at each level`() {
        val lines = listOf(
            "Title",
            "Parent",
            "\tChild",
            "\t\t☑ Done grandchild",
            "\t\t☐ Active grandchild",
        )
        val expected = listOf(
            "Title",
            "Parent",
            "\tChild",
            "\t\t☐ Active grandchild",
            "\t\t☑ Done grandchild",
        )
        assertEquals(expected, CompletedLineUtils.sortCompletedToBottom(lines))
    }

    // ==================== sortCompletedToBottomIndexed ====================

    @Test
    fun `indexed empty input returns empty`() {
        assertEquals(emptyList<Int>(), CompletedLineUtils.sortCompletedToBottomIndexed(emptyList()))
    }

    @Test
    fun `indexed single line returns identity`() {
        assertEquals(listOf(0), CompletedLineUtils.sortCompletedToBottomIndexed(listOf("Title")))
    }

    @Test
    fun `indexed no checked lines returns identity`() {
        val lines = listOf("Title", "☐ A", "☐ B")
        assertEquals(listOf(0, 1, 2), CompletedLineUtils.sortCompletedToBottomIndexed(lines))
    }

    @Test
    fun `indexed checked lines move to bottom`() {
        val lines = listOf("Title", "☑ Done", "☐ Active")
        // "☐ Active" (orig 2) moves to position 1, "☑ Done" (orig 1) moves to position 2
        assertEquals(listOf(0, 2, 1), CompletedLineUtils.sortCompletedToBottomIndexed(lines))
    }

    @Test
    fun `indexed multiple checked and unchecked`() {
        val lines = listOf("Title", "☑ Done A", "☐ Active B", "☐ Active C", "☑ Done D")
        // Unchecked first: Active B (2), Active C (3); then checked: Done A (1), Done D (4)
        assertEquals(listOf(0, 2, 3, 1, 4), CompletedLineUtils.sortCompletedToBottomIndexed(lines))
    }

    @Test
    fun `indexed checked subtrees stay intact`() {
        val lines = listOf("Title", "☑ Done", "\tChild", "☐ Active")
        // Active (3) first, then Done (1) + Child (2)
        assertEquals(listOf(0, 3, 1, 2), CompletedLineUtils.sortCompletedToBottomIndexed(lines))
    }

    @Test
    fun `indexed nested levels sort independently`() {
        val lines = listOf("Title", "Parent", "\t☑ Done child", "\t☐ Active child")
        // Title (0), Parent (1), Active child (3), Done child (2)
        assertEquals(listOf(0, 1, 3, 2), CompletedLineUtils.sortCompletedToBottomIndexed(lines))
    }

    @Test
    fun `indexed empty lines act as barriers`() {
        val lines = listOf("Title", "☑ Done A", "☐ Active B", "", "☑ Done C", "☐ Active D")
        // Section 1: Active B (2), Done A (1); barrier (3); Section 2: Active D (5), Done C (4)
        assertEquals(listOf(0, 2, 1, 3, 5, 4), CompletedLineUtils.sortCompletedToBottomIndexed(lines))
    }

    @Test
    fun `indexed permutation produces same text as sortCompletedToBottom`() {
        val lines = listOf("Title", "☑ Done A", "☐ Active B", "Parent", "\t☑ Child done", "\t☐ Child active")
        val sortedTexts = CompletedLineUtils.sortCompletedToBottom(lines)
        val permutation = CompletedLineUtils.sortCompletedToBottomIndexed(lines)
        val textsFromPermutation = permutation.map { lines[it] }
        assertEquals(sortedTexts, textsFromPermutation)
    }

    // ==================== nearestVisibleLine ====================

    @Test
    fun `nearestVisibleLine returns same index if not hidden`() {
        assertEquals(2, CompletedLineUtils.nearestVisibleLine(listOf("a", "b", "c"), 2, emptySet()))
    }

    @Test
    fun `nearestVisibleLine prefers line above`() {
        val lines = listOf("Title", "Visible", "☑ Hidden", "Below")
        val hidden = setOf(2)
        assertEquals(1, CompletedLineUtils.nearestVisibleLine(lines, 2, hidden))
    }

    @Test
    fun `nearestVisibleLine falls back to line below`() {
        val lines = listOf("☑ Hidden", "☑ Hidden2", "Visible")
        val hidden = setOf(0, 1)
        assertEquals(2, CompletedLineUtils.nearestVisibleLine(lines, 0, hidden))
    }

    @Test
    fun `nearestVisibleLine returns 0 when all hidden`() {
        val lines = listOf("a", "b")
        val hidden = setOf(0, 1)
        assertEquals(0, CompletedLineUtils.nearestVisibleLine(lines, 1, hidden))
    }
}

File: web/src/__tests__/editor/CompletedLineUtils.test.ts
import { describe, it, expect } from 'vitest'
import {
  isCheckedCheckbox,
  computeHiddenIndices,
  computeDisplayItems,
  sortCompletedToBottom,
  nearestVisibleLine,
} from '../../editor/CompletedLineUtils'

describe('isCheckedCheckbox', () => {
  it('returns true for checked line', () => {
    expect(isCheckedCheckbox('☑ Done')).toBe(true)
  })

  it('returns true for indented checked line', () => {
    expect(isCheckedCheckbox('\t\t☑ Done')).toBe(true)
  })

  it('returns false for unchecked line', () => {
    expect(isCheckedCheckbox('☐ Todo')).toBe(false)
  })

  it('returns false for plain line', () => {
    expect(isCheckedCheckbox('Plain text')).toBe(false)
  })

  it('returns false for empty line', () => {
    expect(isCheckedCheckbox('')).toBe(false)
  })
})

describe('computeHiddenIndices', () => {
  it('showCompleted true returns empty set', () => {
    const lines = ['Title', '☑ Done', '☐ Active']
    expect(computeHiddenIndices(lines, true)).toEqual(new Set())
  })

  it('no checked lines returns empty set', () => {
    const lines = ['Title', '☐ Active', '• Bullet']
    expect(computeHiddenIndices(lines, false)).toEqual(new Set())
  })

  it('checked line is hidden', () => {
    const lines = ['Title', '☑ Done', '☐ Active']
    expect(computeHiddenIndices(lines, false)).toEqual(new Set([1]))
  })

  it('checked line with children hides entire subtree', () => {
    const lines = ['Title', '☑ Done', '\tChild 1', '\tChild 2', '☐ Active']
    expect(computeHiddenIndices(lines, false)).toEqual(new Set([1, 2, 3]))
  })

  it('nested checked hides subtree', () => {
    const lines = ['Title', 'Parent', '\t☑ Done', '\t\tGrandchild', '\tSibling']
    expect(computeHiddenIndices(lines, false)).toEqual(new Set([2, 3]))
  })

  it('title line is never hidden even if checked', () => {
    const lines = ['☑ Checked Title', '☐ Active']
    expect(computeHiddenIndices(lines, false)).toEqual(new Set())
  })

  it('multiple checked lines with children', () => {
    const lines = ['Title', '☑ Done1', '\tChild1', '☑ Done2', '\tChild2', '☐ Active']
    expect(computeHiddenIndices(lines, false)).toEqual(new Set([1, 2, 3, 4]))
  })
})

describe('computeDisplayItems', () => {
  it('showCompleted true returns all visible lines', () => {
    const lines = ['Title', '☑ Done', '☐ Active']
    const items = computeDisplayItems(lines, true)
    expect(items).toEqual([
      { type: 'visible', realIndex: 0 },
      { type: 'visible', realIndex: 1 },
      { type: 'visible', realIndex: 2 },
    ])
  })

  it('hidden lines produce placeholder with correct count', () => {
    const lines = ['Title', '☑ Done1', '☑ Done2', '☐ Active']
    const items = computeDisplayItems(lines, false)
    expect(items).toEqual([
      { type: 'visible', realIndex: 0 },
      { type: 'placeholder', count: 2, indentLevel: 0, startIndex: 1, endIndex: 2 },
      { type: 'visible', realIndex: 3 },
    ])
  })

  it('placeholder counts only top-level hidden items', () => {
    const lines = ['Title', '☑ Done', '\tChild1', '\tChild2', '☐ Active']
    const items = computeDisplayItems(lines, false)
    expect(items).toEqual([
      { type: 'visible', realIndex: 0 },
      { type: 'placeholder', count: 1, indentLevel: 0, startIndex: 1, endIndex: 3 },
      { type: 'visible', realIndex: 4 },
    ])
  })

  it('indented placeholder has correct indent level', () => {
    const lines = ['Title', 'Parent', '\t☑ Done', '\t☐ Active']
    const items = computeDisplayItems(lines, false)
    expect(items).toEqual([
      { type: 'visible', realIndex: 0 },
      { type: 'visible', realIndex: 1 },
      { type: 'placeholder', count: 1, indentLevel: 1, startIndex: 2, endIndex: 2 },
      { type: 'visible', realIndex: 3 },
    ])
  })

})

describe('sortCompletedToBottom', () => {
  it('empty input returns empty', () => {
    expect(sortCompletedToBottom([])).toEqual([])
  })

  it('single line returns unchanged', () => {
    expect(sortCompletedToBottom(['Title'])).toEqual(['Title'])
  })

  it('no checked lines returns unchanged', () => {
    const lines = ['Title', '☐ A', '☐ B']
    expect(sortCompletedToBottom(lines)).toEqual(lines)
  })

  it('checked lines move to bottom', () => {
    const lines = ['Title', '☑ Done', '☐ Active']
    expect(sortCompletedToBottom(lines)).toEqual(['Title', '☐ Active', '☑ Done'])
  })

  it('multiple checked and unchecked sort correctly', () => {
    const lines = ['Title', '☑ Done A', '☐ Active B', '☐ Active C', '☑ Done D']
    expect(sortCompletedToBottom(lines)).toEqual([
      'Title',
      '☐ Active B',
      '☐ Active C',
      '☑ Done A',
      '☑ Done D',
    ])
  })

  it('checked subtrees stay intact', () => {
    const lines = ['Title', '☑ Done', '\tChild', '☐ Active']
    expect(sortCompletedToBottom(lines)).toEqual(['Title', '☐ Active', '☑ Done', '\tChild'])
  })

  it('nested levels sort independently', () => {
    const lines = ['Title', 'Parent', '\t☑ Done child', '\t☐ Active child']
    expect(sortCompletedToBottom(lines)).toEqual([
      'Title',
      'Parent',
      '\t☐ Active child',
      '\t☑ Done child',
    ])
  })

  it('empty lines act as barriers', () => {
    const lines = ['Title', '☑ Done A', '☐ Active B', '', '☑ Done C', '☐ Active D']
    expect(sortCompletedToBottom(lines)).toEqual([
      'Title',
      '☐ Active B',
      '☑ Done A',
      '',
      '☐ Active D',
      '☑ Done C',
    ])
  })

  it('already sorted is idempotent', () => {
    const lines = ['Title', '☐ Active', '☑ Done']
    expect(sortCompletedToBottom(lines)).toEqual(lines)
  })

  it('deep nesting sorts at each level', () => {
    const lines = [
      'Title',
      'Parent',
      '\tChild',
      '\t\t☑ Done grandchild',
      '\t\t☐ Active grandchild',
    ]
    expect(sortCompletedToBottom(lines)).toEqual([
      'Title',
      'Parent',
      '\tChild',
      '\t\t☐ Active grandchild',
      '\t\t☑ Done grandchild',
    ])
  })
})

describe('nearestVisibleLine', () => {
  it('returns same index if not hidden', () => {
    expect(nearestVisibleLine(['a', 'b', 'c'], 2, new Set())).toBe(2)
  })

  it('prefers line above', () => {
    const lines = ['Title', 'Visible', '☑ Hidden', 'Below']
    expect(nearestVisibleLine(lines, 2, new Set([2]))).toBe(1)
  })

  it('falls back to line below', () => {
    const lines = ['☑ Hidden', '☑ Hidden2', 'Visible']
    expect(nearestVisibleLine(lines, 0, new Set([0, 1]))).toBe(2)
  })

  it('returns 0 when all hidden', () => {
    const lines = ['a', 'b']
    expect(nearestVisibleLine(lines, 1, new Set([0, 1]))).toBe(0)
  })
})
